pkg/operator/controller/ingress/status.go

```diff
@@ -171,35 +182,108 @@ func computeDeploymentPodsScheduledCondition(deployment *appsv1.Deployment, pods
 }
 
 // computeIngressAvailableCondition computes the ingress controller's current Available status state
-// by inspecting the Available condition of deployment. The ingresscontroller is only available if
-// the deployment is also available.
-func computeIngressAvailableCondition(deployment *appsv1.Deployment) operatorv1.OperatorCondition {
-	for _, cond := range deployment.Status.Conditions {
-		if cond.Type != appsv1.DeploymentAvailable {
+// by inspecting the following:
+// 1) the Available condition of Deployment,
+// 2) the DNSReady condition of the IngressController, and
+// 3) the LoadBalancerReady condition of the IngressController.
+// The ingresscontroller is judged Available only if all 3 conditions are true
+func computeIngressAvailableCondition(conditions []operatorv1.OperatorCondition) operatorv1.OperatorCondition {
+
+	expected := []expectedCondition{
+		{
+			condition: IngressControllerDeploymentAvailableConditionType,
+			status:    operatorv1.ConditionTrue,
+		},
+		{
+			condition: operatorv1.DNSReadyIngressConditionType,
+			status:    operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{
+				operatorv1.LoadBalancerManagedIngressConditionType,
+				operatorv1.LoadBalancerReadyIngressConditionType,
+				operatorv1.DNSManagedIngressConditionType,
+			},
+			gracePeriod: time.Second * 30,
+		},
+		{
+			condition:        operatorv1.LoadBalancerReadyIngressConditionType,
+			status:           operatorv1.ConditionTrue,
+			ifConditionsTrue: []string{operatorv1.LoadBalancerManagedIngressConditionType},
+			gracePeriod:      time.Second * 90,
+		},
+	}
+
+	// Cover the rare case of no conditions
+	if len(conditions) == 0 {
+		return operatorv1.OperatorCondition{Type: operatorv1.OperatorStatusTypeAvailable, Status: operatorv1.ConditionFalse}
+	}
```

Comment:
`conditions` should never be empty (`syncIngressControllerStatus` sets a whole bunch of status conditions before it calls `computeIngressAvailableCondition`).  Is this check only needed to make the "all availability not present" unit test case pass, or is the intention to check if none of the conditions mentioned in `expected` is present?

Reply:
Yes, it was added to make the unit test case pass.  But also, conditions don't need to be checked if there are none, even if that is extremely unlikely.